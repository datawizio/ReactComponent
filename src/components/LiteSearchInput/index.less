@import "../../styles/variables";
@import "./rtl";

.lite-search {
  width: 100%;
  display: block;
  position: relative;

  &__search-icon {
    top: 4px;
    position: absolute;

    font-size: 15px;
    color: @primary-color;
  }

  &__cancel-icon {
    top: 5px;
    right: 0;
    position: absolute;

    font-size: 12px;
    color: @primary-color;
  }

  &__input {
    background-color: @input-bg-color;

    &:focus {
      outline: none;
      border-bottom: 1px solid @primary-color;
    }

    width: 100%;
    border: none;
    padding-left: 20px;
    padding-right: 15px;
    border-bottom: 1px solid #b3b1b1;
  }
}
