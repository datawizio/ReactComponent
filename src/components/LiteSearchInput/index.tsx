import * as React from "react";

import { useDebouncedCallback } from "use-debounce";
import { useState, useCallback, useEffect } from "react";
import { SearchOutlined, CloseOutlined } from "@ant-design/icons";

import "./index.less";

export interface LiteSearchInputProps
  extends React.InputHTMLAttributes<HTMLInputElement> {
  /**
   * Задержка вызова onSearch
   */
  debounceDelay?: number;

  /**
   * Вызываеться при нажатии на крестик
   */
  onClear?: () => void;

  /**
   * Работает как OnChange с учётом задержки debounceDelay
   */
  onSearch?: (value: string) => void;
}

const LiteSearchInput: React.FC<LiteSearchInputProps> = props => {
  const {
    style,
    onClear,
    onSearch,
    onChange,
    debounceDelay,
    ...restProps
  } = props;

  const [value, setValue] = useState(() => props.value || "");

  const [onSearchDebounced] = useDebouncedCallback((value: string) => {
    onSearch && onSearch(value);
  }, debounceDelay);

  const handleChange = useCallback(
    event => {
      setValue(event.target.value);
      // run callbacks
      onChange && onChange(event);
      onSearchDebounced(event.target.value);
    },
    [onChange, onSearchDebounced]
  );

  const handleClear = useCallback(() => {
    setValue("");
    onClear && onClear();
    onSearchDebounced("");
  }, [onClear, onSearchDebounced]);

  useEffect(() => {
    setValue(props.value ? props.value : "");
  }, [props.value]);

  return (
    <span className="lite-search" style={style}>
      <SearchOutlined className="lite-search__search-icon" />
      <input
        {...restProps}
        value={value}
        onChange={handleChange}
        className="lite-search__input"
      />
      {value && (
        <CloseOutlined
          onClick={handleClear}
          className="lite-search__cancel-icon"
        />
      )}
    </span>
  );
};

LiteSearchInput.defaultProps = {
  debounceDelay: 0
};

export default LiteSearchInput;
