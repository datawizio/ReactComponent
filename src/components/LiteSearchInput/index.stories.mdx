import { useState } from "react";
import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import LiteSearchInput from "./index";

<Meta title="Components/Input/LiteSearchIInput" component={LiteSearchInput} />

# LiteSearchInput

Компонент предназначен для ввода текста для поиска.

## Basic

<Preview>
  <Story name="Basic">
    <LiteSearchInput />
  </Story>
</Preview>

## With debounce

<Preview>
  <Story name="With debounce">
    {() => {
      const [value, setValue] = useState("");
      //
      return (
        <div>
          <LiteSearchInput
            debounceDelay={1000}
            onSearch={value => setValue(value)}
          />
          <span>Search text: {value}</span>
        </div>
      );
    }}
  </Story>
</Preview>

## Props

<Props of={LiteSearchInput} />
