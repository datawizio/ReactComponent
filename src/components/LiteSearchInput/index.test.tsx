import "jsdom-global/register";
import React from "react";
import { mount } from "enzyme";

import LiteSearchInput from "./index";

const mockProps = {
  style: {},
  onClear: jest.fn(),
  onSearch: jest.fn(),
  onChange: jest.fn(),
  debounceDelay: 1
};

const setUp = (props?) => mount(<LiteSearchInput {...props} />);

describe("LiteSearchInput component", () => {
  let component;
  beforeEach(() => {
    component = setUp(mockProps);
  });

  it("LiteSearchInput rendered correctly", () => {
    expect(component).toMatchSnapshot();
  });

  it("LiteSearchInput handleChange works correctly", () => {
    const searchBtn = component.find(".lite-search__input");
    const mockedEvent = {
      target: {
        value: "value"
      }
    };
    searchBtn.simulate("change", mockedEvent);
    expect(component.props().onChange).toBeCalledTimes(1);
  });

  it("LiteSearchInput handleClear works correctly", () => {
    const searchBtn = component.find(".lite-search__input");
    const mockedEvent = {
      target: {
        value: "value"
      }
    };
    component.setProps({ onChange: jest.fn() });
    searchBtn.simulate("change", mockedEvent);
    const clearBtn = component.find(".lite-search__cancel-icon");

    clearBtn.first().simulate("click");
    expect(component.props().onClear).toBeCalledTimes(1);
  });

  it("LiteSearchInput update with empty value", () => {
    component.setProps({ value: "" });
    const searchBtn = component.find(".lite-search__input");

    expect(searchBtn.prop("value")).toBe("");
  });
});
