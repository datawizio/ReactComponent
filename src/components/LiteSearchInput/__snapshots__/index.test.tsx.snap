// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiteSearchInput component LiteSearchInput rendered correctly 1`] = `
<LiteSearchInput
  debounceDelay={1}
  onChange={[MockFunction]}
  onClear={[MockFunction]}
  onSearch={[MockFunction]}
  style={Object {}}
>
  <span
    className="lite-search"
    style={Object {}}
  >
    <ForwardRef(SearchOutlined)
      className="lite-search__search-icon"
    >
      <AntdIcon
        className="lite-search__search-icon"
        icon={
          Object {
            "icon": Object {
              "attrs": Object {
                "focusable": "false",
                "viewBox": "64 64 896 896",
              },
              "children": Array [
                Object {
                  "attrs": Object {
                    "d": "M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z",
                  },
                  "tag": "path",
                },
              ],
              "tag": "svg",
            },
            "name": "search",
            "theme": "outlined",
          }
        }
      >
        <span
          aria-label="search"
          className="anticon anticon-search lite-search__search-icon"
          role="img"
        >
          <IconReact
            className=""
            icon={
              Object {
                "icon": Object {
                  "attrs": Object {
                    "focusable": "false",
                    "viewBox": "64 64 896 896",
                  },
                  "children": Array [
                    Object {
                      "attrs": Object {
                        "d": "M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z",
                      },
                      "tag": "path",
                    },
                  ],
                  "tag": "svg",
                },
                "name": "search",
                "theme": "outlined",
              }
            }
          >
            <svg
              aria-hidden="true"
              className=""
              data-icon="search"
              fill="currentColor"
              focusable="false"
              height="1em"
              key="svg-search"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                key="svg-search-svg-0"
              />
            </svg>
          </IconReact>
        </span>
      </AntdIcon>
    </ForwardRef(SearchOutlined)>
    <input
      className="lite-search__input"
      onChange={[Function]}
      value=""
    />
  </span>
</LiteSearchInput>
`;
